roofit/codegen/inc/RooFit/CodegenImpl.h

```diff
@@ -0,0 +1,155 @@
+void codegenImpl(RooRealSumFunc &arg, CodegenContext &ctx);
+void codegenImpl(RooRealSumPdf &arg, CodegenContext &ctx);
+void codegenImpl(RooRealVar &arg, CodegenContext &ctx);
+void codegenImpl(RooRecursiveFraction &arg, CodegenContext &ctx);
+void codegenImpl(RooStats::HistFactory::FlexibleInterpVar &arg, CodegenContext &ctx);
+void codegenImpl(RooUniform &arg, CodegenContext &ctx);
+
+std::string codegenIntegralImpl(RooAbsReal &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooBernstein &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooBifurGauss &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooCBShape &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooChebychev &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooEfficiency &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooExponential &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooGamma &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooGaussian &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooHistFunc &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooHistPdf &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooLandau &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooLognormal &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooMultiVarGaussian &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooPoisson &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooPolyVar &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooPolynomial &arg, int code, const char *rangeName, CodegenContext &ctx);
+std::string codegenIntegralImpl(RooUniform &arg, int code, const char *rangeName, CodegenContext &ctx);
+
+template <class Arg_t, int P>
+std::string codegenIntegralImpl(Arg_t &arg, int code, const char *rangeName, CodegenContext &ctx, Prio<P> p)
+{
+   if constexpr (std::is_same<Prio<P>, PrioLowest>::value) {
+      return codegenIntegralImpl(arg, code, rangeName, ctx);
+   } else {
+      return codegenIntegralImpl(arg, code, rangeName, ctx, p.next());
+   }
+}
+
+template <class Arg_t>
+struct CodegenIntegralImplCaller {
+
+   static auto call(RooAbsReal &arg, int code, const char *rangeName, RooFit::CodegenContext &ctx)
+   {
+      return codegenIntegralImpl(static_cast<Arg_t &>(arg), code, rangeName, ctx, PrioHighest{});
+   }
+};
```

Comment:
Do I understand correctly that the Caller first tries to call the prio 1 (highest) implementation, and then will loop through all 10 using the Impl template until it finds one that is actually implemented for the Arg_t already so doesn't need to use the template?

Few questions:
1. Do you actually need all 10 priority levels? How about replacing that with just an overload boolean? It will save some looping through empty functions.
2. I'm not sure, but there may be some pathological situations. For instance, what if in an interactive session you call some function, the tree of templates gets instantiated and then afterwards you try to define a new overload? Won't that overload then already have been instantiated from the Impl template? Another possible situation would be something similar to the above but in separate compilation units.